### ISSUE TITLE
Chinese language support

### ISSUE THREAD

**Issue by inVains** [digininja/DVWA#432]
Thanks for the DVWA!  It help me to learn and it's really good for beginers.
Expecting Chinese language support, as many Chinese students use it with poor English.

Need Chinese language support For:
1. README.md
2. Contents in pages, including help Pages.
3. Code comments.

Planning to implement it and Pull.
Is it encouraged ?

**Comment by digininja** [comment]
I'll start with yes, all improvements are welcome and I'd love to see this be accessible by a larger audience.

But, I'll add these caveats:

* As I don't know you, and you don't have any GitHub reputation, at least at the start I'd need to find someone to fully check what you've written. I can't blindly trust you've done an honest job.
* What you change can't affect the normal use of the site by English speakers, so we would need to find a way to put the translations in so they are switchable with English or in a separate file. For the README that would be easy, it can be in README.cn.md or something like that..
* You would need to own the translations and keep them up to date when other things change. That isn't very often, but I don't want people coming to me complaining instructions don't work because they are a version old.

I'd also like to see how vim handles a page with Chinese characters in it. I've worked on sites with non-ASCII before and sometimes editing the files is OK and sometimes it completely ruins them.

What I'd suggest is to start with the README, that will probably have the biggest impact as it will help people get it installed and working. That will let us see how things fit together. I'll have a think about how it could be incorporated into things like the help files.

So definitely interested, I just want to make sure it is done right from the start rather than rushing in and then having to back things out because they break stuff.

**Comment by inVains** [comment]
Your caveats and concern is reasonable.
I'll start as you suggestion:).

**Comment by songzy12** [comment]
Did a quick check about this. This seems the topic "multi-language support for php sites".

Specific to the DVWA project, and the contents mentioned above, my personal opinion is README and help pages are both good candidates for translation, given they:
1. are manily descriptive sentences written in natrual languages
2. contains helpful info which are worth reading even without the coding challenges/practices
3. are stored seperatly which makes them easy to maintain.

For the RAEDME, we already translated and reviewed the README.cn.md.

For the help pages, my humble suggestion is:
1. add a new config option like `$_DVWA[ 'default_locale' ] = 'en';` in `config/config.inc.php`
2. then in function `dvwaButtonHelpHtmlGet` of `dvwa/includes/dvwaPage.inc.php`, we get this config and add the parameter `locale` when fetching page `vulnerabilities/view_help.php?id={$pId}&security={$security}&locale={$locale}`
3. in file `vulnerabilities/view_help.php`, if `{$locale} != 'en'`, we get `vulnerabilities/{$id}/help/help.{$locale}.php`.
4. the specific translated help pages can be organized as, take brute force as an example, `vulnerabilities/brute/help/help.zh.php` along with `vulnerabilities/brute/help/help.php`.

If the above plan looks good, I can try to implement the feature as decripted in above step 1,2,3. 

Once the feature is ready, volunteers (not only for Chinese) can work on tranlations, e.g., 
1. in a seperate git branch to not mess up the master with incomplete contents and for easy review. 
2. when all translations are finished, we can merge the seperate branches into master.

We can use ISO 639-1 language codes [1, 2], i.e., `en` for English and `zh` for Chinese.

[1] https://en.wikipedia.org/wiki/ISO_639-1
[2] https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes

**Comment by inVains** [comment]
Agree with it: README and help pages should be enough for learners. 
And they are nessesary to undestand the basic English and terminologies in English.

I'll do what I can help.

> Did a quick check about this. This seems the topic "multi-language support for php sites".
> 
> Specific to the DVWA project, and the contents mentioned above, my personal opinion is README and help pages are both good candidates for translation, given they:
> 
> 1. are manily descriptive sentences written in natrual languages
> 2. contains helpful info which are worth reading even without the coding challenges/practices
> 3. are stored seperatly which makes them easy to maintain.
> 
> For the RAEDME, we already translated and reviewed the README.cn.md.
> 
> For the help pages, my humble suggestion is:
> 
> 1. add a new config option like `$_DVWA[ 'default_locale' ] = 'en';` in `config/config.inc.php`
> 2. then in function `dvwaButtonHelpHtmlGet` of `dvwa/includes/dvwaPage.inc.php`, we get this config and add the parameter `locale` when fetching page `vulnerabilities/view_help.php?id={$pId}&security={$security}&locale={$locale}`
> 3. in file `vulnerabilities/view_help.php`, if `{$locale} != 'en'`, we get `vulnerabilities/{$id}/help/help.{$locale}.php`.
> 4. the specific translated help pages can be organized as, take brute force as an example, `vulnerabilities/brute/help/help.zh.php` along with `vulnerabilities/brute/help/help.php`.
> 
> If the above plan looks good, I can try to implement the feature as decripted in above step 1,2,3.
> 
> Once the feature is ready, volunteers (not only for Chinese) can work on tranlations, e.g.,
> 
> 1. in a seperate git branch to not mess up the master with incomplete contents and for easy review.
> 2. when all translations are finished, we can merge the seperate branches into master.
> 
> We can use ISO 639-1 language codes [1, 2], i.e., `en` for English and `zh` for Chinese.
> 
> [1] https://en.wikipedia.org/wiki/ISO_639-1
> [2] https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes

**Comment by digininja** [comment]
Seems like a good plan to me.

Feel free to rename the README from ch to zh if that is how it is supposed
to be, I just assumed ch, now I think about it, that is probably
Switzerland.

I wonder if we could get a vulnerability in here, maybe set it in the
config file but let it be overridden in a cookie and then turn that into a
LFI vulnerability. Do it as a normal build first and we could look at that
later.

**Comment by Basyaact** [comment]
HI,if you don't mind ,check my Full-Chinese version DVWA in https://github.com/Basyaact/DVWA-Chinese

**Comment by digininja** [comment]
@Basyaact I like the enthusiasm, but trying to keep this up to date with the live version isn't going to be easy doing it the way you've done it. If you look at what is being done as part of this bit of work, the others are looking at putting an in-app translation together so you can select a language to use rather than having a completely standalone version.

@inVains and @songzy12 sorry, I got distracted from this, I'll make sure I get a look at it this week to get things moving again.

**Comment by Basyaact** [comment]
@digininja I see, I will keep this Chinese version up-to-date in time for Chinese user ,until got a Official Chinese release.
if needed,i can join the translate team too

**Comment by digininja** [comment]
Have a read through what has been done here so far and feel free to join in.

The general plan is to have a switch somewhere which lets you chose your
language which is then loaded from a config file or hardcoded in some way.
In the long term it would be good to completely extract all the human
readable text from the app and have it all in resource files to make it
easy to add new languages while not affecting the app itself.

**Comment by Basyaact** [comment]
Got it, my primary language is Chinese and I'd like to join in.Ready in anytime

**Comment by songzy12** [comment]
@inVains @Basyaact 

FYI with the merge of https://github.com/digininja/DVWA/pull/438, the support for translation of help pages of vulnerabilities is ready: 
![1628961001(1)](https://user-images.githubusercontent.com/8758315/129455042-a5af4a01-d462-444a-87c9-ab7a5189bc5e.png)

How to verify zh help pages: 
1. set `$_DVWA[ 'default_locale' ] = 'zh';` in `config/config.inc.php`: https://github.com/digininja/DVWA/blob/2d62b7b8ca19499dee6f08c727ab0692725443ff/config/config.inc.php.dist#L48
2. add translated help pages in coresponding folder, e.g., add translated `vulnerabilities/brute/help/help.zh.php` besides the original `vulnerabilities/brute/help/help.php`.
3. then click the "view help" button on the vulnerability page, you will see the content of new added page.

How to proceed with the Chinese language support:
1. I would suggest to add **one** help page for **one** pull request, and each pull request only contains **one** commit, to keep the git history tidy and clean.
2. when there are pull requests regarding translated help pages, I can help with the review process. 

Future work:
Now we are clear with how the help pages for vulnerabilities can be supported for multilingual, 
how to expand this to other tabs (like "Home", "Instructions", "Setup/Reset DB") still needs some investigation.
![1628960726(1)](https://user-images.githubusercontent.com/8758315/129454912-cc4ade06-c7b1-4811-8c9b-1c3b76bb3981.png)

**Comment by digininja** [comment]
@songzy12  write up, thanks.

I'm about to do some work on SQLi and blind SQLi to add some new features, so best start with something else till I get that stabilised. I'm only adding new things so anything that has been done won't be lost, it will just need extra adding to it.

**Comment by cclin77** [comment]
+1, thingk
